src/tests/suite/daos_container.c

```diff
@@ -771,6 +771,226 @@ co_set_prop(void **state)
+	 * is an alias for create+del container.
+	 */
+	pool_prop = get_daos_prop_with_owner_acl_perms(DAOS_ACL_PERM_POOL_ALL &
+						       ~DAOS_ACL_PERM_DEL_CONT &
+						       ~DAOS_ACL_PERM_WRITE,
+						       DAOS_PROP_PO_ACL);
+
+	/* container doesn't give delete privs to the owner */
+	cont_prop = get_daos_prop_with_owner_acl_perms(DAOS_ACL_PERM_CONT_ALL &
+						       ~DAOS_ACL_PERM_DEL_CONT,
+						       DAOS_PROP_CO_ACL);
+
+	while (!rc && arg->setup_state != SETUP_CONT_CREATE)
+		rc = test_setup_next_step((void **)&arg, NULL, pool_prop,
+					  cont_prop);
+	assert_int_equal(rc, 0);
+
+	MPI_Barrier(MPI_COMM_WORLD);
+
+	rc = daos_cont_destroy(arg->pool.poh, arg->co_uuid, 1, NULL);
+	assert_int_equal(rc, -DER_NO_PERM);
+
+	print_message("Delete with privs from container ACL only\n");
+
+	cont_acl = daos_acl_dup(cont_prop->dpp_entries[0].dpe_val_ptr);
+	assert_non_null(cont_acl);
+	rc = daos_acl_get_ace_for_principal(cont_acl, DAOS_ACL_OWNER, NULL,
+					    &update_ace);
+	assert_int_equal(rc, 0);
+	update_ace->dae_allow_perms = DAOS_ACL_PERM_CONT_ALL;
+
+	print_message("- getting container handle\n");
+	rc = daos_cont_open(arg->pool.poh, arg->co_uuid, DAOS_COO_RW, &coh,
+			    NULL, NULL);
+	assert_int_equal(rc, 0);
+
+	print_message("- updating cont ACL to restore delete privs\n");
+	rc = daos_cont_update_acl(coh, cont_acl, NULL);
+	assert_int_equal(rc, 0);
+
+	print_message("- closing container\n");
+	rc = daos_cont_close(coh, NULL);
+	assert_int_equal(rc, 0);
+
+	print_message("Deleting container now should succeed\n");
+	rc = daos_cont_destroy(arg->pool.poh, arg->co_uuid, 1, NULL);
+	assert_int_equal(rc, 0);
+
+	MPI_Barrier(MPI_COMM_WORLD);
+
+	/* Clear cont uuid since we already deleted it */
+	uuid_clear(arg->co_uuid);
+
+	daos_acl_free(cont_acl);
+	daos_prop_free(pool_prop);
+	daos_prop_free(cont_prop);
+	test_teardown((void **)&arg);
+}
+
+static void
+co_destroy_allowed_by_pool(void **state)
+{
+	test_arg_t	*arg0 = *state;
+	test_arg_t	*arg = NULL;
+	daos_prop_t	*pool_prop;
+	daos_prop_t	*cont_prop;
+	int		 rc;
+
+	rc = test_setup((void **)&arg, SETUP_EQ, arg0->multi_rank,
+			DEFAULT_POOL_SIZE, NULL);
+	assert_int_equal(rc, 0);
+
+	print_message("Delete a container where only pool allows it\n");
+
+	/* pool gives the owner all privs, including delete cont */
+	pool_prop = get_daos_prop_with_owner_acl_perms(DAOS_ACL_PERM_POOL_ALL,
+						       DAOS_PROP_PO_ACL);
+
+	/* container doesn't give delete privs to the owner */
+	cont_prop = get_daos_prop_with_owner_acl_perms(DAOS_ACL_PERM_CONT_ALL &
+						       ~DAOS_ACL_PERM_DEL_CONT,
+						       DAOS_PROP_CO_ACL);
+
+	while (!rc && arg->setup_state != SETUP_CONT_CREATE)
+		rc = test_setup_next_step((void **)&arg, NULL, pool_prop,
+					  cont_prop);
+	assert_int_equal(rc, 0);
+
+	MPI_Barrier(MPI_COMM_WORLD);
+
+	print_message("Deleting container should succeed\n");
+	rc = daos_cont_destroy(arg->pool.poh, arg->co_uuid, 1, NULL);
```

Comment:
similar question: should this be done by all MPI ranks or just rank 0?

Reply:
I'll switch this to rank 0 only.